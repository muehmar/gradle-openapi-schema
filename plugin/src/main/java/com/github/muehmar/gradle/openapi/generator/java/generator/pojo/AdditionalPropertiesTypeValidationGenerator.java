package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.ValidationGenerator.assertTrue;
import static com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties.additionalPropertiesName;
import static io.github.muehmar.codegenerator.Generator.constant;

import com.github.muehmar.gradle.openapi.generator.java.generator.shared.SettingsFunctions;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.MethodGenBuilder;

public class AdditionalPropertiesTypeValidationGenerator {
  private AdditionalPropertiesTypeValidationGenerator() {}

  public static Generator<JavaObjectPojo, PojoSettings>
      additionalPropertiesTypeValidationGenerator() {
    return annotatedCorrectTypeMethod()
        .filter(JavaAdditionalProperties::isAllowed)
        .filter(JavaAdditionalProperties::isNotValueAnyType)
        .contraMap(JavaObjectPojo::getAdditionalProperties);
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> annotatedCorrectTypeMethod() {
    return JacksonAnnotationGenerator.<JavaAdditionalProperties>jsonIgnore()
        .append(
            assertTrue(
                props ->
                    String.format(
                        "Not all additional properties are instances of %s",
                        props.getType().getQualifiedClassName().getClassName())))
        .append(correctTypeMethod());
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> correctTypeMethod() {
    return MethodGenBuilder.<JavaAdditionalProperties, PojoSettings>create()
        .modifiers(SettingsFunctions::validationMethodModifiers)
        .noGenericTypes()
        .returnType("boolean")
        .methodName("isAllAdditionalPropertiesHaveCorrectType")
        .noArguments()
        .content(
            constant(
                String.format(
                    "return getAdditionalProperties().size() == %s.size();",
                    additionalPropertiesName())))
        .build();
  }
}
