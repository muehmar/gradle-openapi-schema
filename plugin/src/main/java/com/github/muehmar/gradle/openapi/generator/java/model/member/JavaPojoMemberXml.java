package com.github.muehmar.gradle.openapi.generator.java.model.member;

import com.github.muehmar.gradle.openapi.generator.model.PojoMemberXml;
import java.util.Optional;
import lombok.Value;

@Value
public class JavaPojoMemberXml {
  Optional<String> name;
  Optional<Boolean> isAttribute;

  public static JavaPojoMemberXml noDefinition() {
    return new JavaPojoMemberXml(Optional.empty(), Optional.empty());
  }

  public static JavaPojoMemberXml fromPojoMemberXml(PojoMemberXml pojoMemberXml) {
    return new JavaPojoMemberXml(pojoMemberXml.getName(), pojoMemberXml.getIsAttribute());
  }

  public boolean hasDefinition() {
    return name.isPresent() || isAttribute.isPresent();
  }
}
