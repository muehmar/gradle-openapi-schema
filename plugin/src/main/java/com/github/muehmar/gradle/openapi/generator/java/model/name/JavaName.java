package com.github.muehmar.gradle.openapi.generator.java.model.name;

import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import lombok.EqualsAndHashCode;

/**
 * Name for Java identifiers, like member, method or classnames. A {@link JavaName} can be
 * constructed from any string and gets escaped to conform to the specification for identifiers.
 * Render the name as string with {@link JavaName#asString()} or {@link JavaName#toString()} will
 * result in the escaped string.<br>
 * <br>
 * Any operation like {@link JavaName#prefix(String)} is performed on the original string rather
 * than on the escaped string. This will produce identifiers which are as close as possible to the
 * original name, i.e. only do escaping if necessary. E.g. if a Java keyword is used as name and
 * gets prefixed, there is no longer the need to escape it.
 */
@EqualsAndHashCode
public class JavaName {
  private final Name originalName;
  private final JavaIdentifier identifier;

  private JavaName(Name name) {
    this.originalName = name;
    this.identifier = JavaIdentifier.fromName(name);
  }

  public static JavaName fromName(Name name) {
    return new JavaName(name);
  }

  public static JavaName fromString(String name) {
    return new JavaName(Name.ofString(name));
  }

  public JavaName prefix(String prefix) {
    return new JavaName(originalName.prefix(prefix));
  }

  public JavaName append(String append) {
    return new JavaName(originalName.append(append));
  }

  public JavaName startUpperCase() {
    return new JavaName(originalName.startUpperCase());
  }

  public JavaName startLowerCase() {
    return new JavaName(originalName.startLowerCase());
  }

  public JavaName prefixedMethodName(String prefix) {
    if (prefix.isEmpty()) {
      return startLowerCase();
    } else {
      return startUpperCase().prefix(prefix);
    }
  }

  public String wordBoundaryPattern() {
    return String.format("\\b%s\\b", identifier);
  }

  public Name getOriginalName() {
    return originalName;
  }

  public String asString() {
    return identifier.asString();
  }

  @Override
  public String toString() {
    return asString();
  }
}
