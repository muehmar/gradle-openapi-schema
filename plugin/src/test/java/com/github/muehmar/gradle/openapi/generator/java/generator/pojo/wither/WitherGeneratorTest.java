package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.wither;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.wither.WitherGenerator.witherGenerator;
import static com.github.muehmar.gradle.openapi.generator.java.model.member.JavaPojoMemberBuilder.javaPojoMemberBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.model.member.TestJavaPojoMembers.requiredBirthdate;
import static com.github.muehmar.gradle.openapi.generator.java.model.member.TestJavaPojoMembers.requiredString;
import static com.github.muehmar.gradle.openapi.generator.java.model.name.JavaPojoNames.invoiceName;
import static com.github.muehmar.gradle.openapi.generator.java.model.type.JavaTypes.stringType;
import static com.github.muehmar.gradle.openapi.generator.model.Nullability.NULLABLE;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil.writerSnapshot;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.member.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.member.JavaPojoMemberXml;
import com.github.muehmar.gradle.openapi.generator.java.model.member.JavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import java.util.stream.Stream;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

@SnapshotTest
class WitherGeneratorTest {
  private Expect expect;

  @ParameterizedTest
  @MethodSource("allNullabilityAndNecessityVariantsSingleMember")
  @SnapshotName("allNullabilityAndNecessityVariants")
  void generate_when_calledWithNullabilityAndNecessityVariants_then_correctOutput(
      WitherGenerator.WitherContent witherContent, JavaPojoMember member) {
    final Generator<WitherGenerator.WitherContent, PojoSettings> generator = witherGenerator();

    final WitherGenerator.WitherContent singleMemberWitherContent =
        WitherContentBuilder.fullWitherContentBuilder()
            .className(witherContent.getClassName())
            .membersForWithers(PList.single(member))
            .technicalPojoMembers(witherContent.getTechnicalPojoMembers())
            .build();

    final Writer writer =
        generator.generate(singleMemberWitherContent, defaultTestSettings(), javaWriter());

    expect.scenario(member.getName().asString()).toMatchSnapshot(writerSnapshot(writer));
  }

  static Stream<Arguments> allNullabilityAndNecessityVariantsSingleMember() {
    final JavaObjectPojo pojo = JavaPojos.allNecessityAndNullabilityVariants();
    final WitherGenerator.WitherContent witherContent = pojo.getWitherContent();
    final JavaObjectPojo pojoTypeMapped = JavaPojos.allNecessityAndNullabilityVariantsTypeMapped();
    final JavaObjectPojo pojoTypeMappedWithCorrectNames =
        pojoTypeMapped.withMembers(
            JavaPojoMembers.fromMembers(
                pojoTypeMapped
                    .getMembers()
                    .map(member -> member.withName(member.getName().append("Mapped")))));
    final WitherGenerator.WitherContent witherContentTypeMapped =
        pojoTypeMappedWithCorrectNames.getWitherContent();
    return pojo.getMembers()
        .map(member -> Arguments.of(witherContent, member))
        .concat(
            pojoTypeMappedWithCorrectNames
                .getMembers()
                .map(member -> Arguments.of(witherContentTypeMapped, member)))
        .toStream();
  }

  @Test
  @SnapshotName("noAdditionalProperties")
  void generate_when_noAdditionalProperties_then_correctOutput() {
    final Generator<WitherGenerator.WitherContent, PojoSettings> generator = witherGenerator();
    final Writer writer =
        generator.generate(
            WitherContentBuilder.create()
                .className(JavaName.fromString("ObjectDto"))
                .membersForWithers(PList.single(requiredString()))
                .technicalPojoMembers(
                    PList.of(requiredString(), requiredBirthdate())
                        .flatMap(JavaPojoMember::getTechnicalMembers))
                .build(),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("illegalIdentifierPojo")
  void generate_when_illegalIdentifierPojo_then_correctOutput() {
    final Generator<WitherGenerator.WitherContent, PojoSettings> generator = witherGenerator();

    final Writer writer =
        generator.generate(
            JavaPojos.illegalIdentifierPojo().getWitherContent(),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("propertyNameMatchesSubstringOfOtherProperty")
  void
      generate_when_pojoContainsPropertyNameWhichMatchesSubstringOfOtherProperty_then_correctOutput() {
    final JavaPojoMember surnameMember =
        javaPojoMemberBuilder()
            .pojoName(invoiceName())
            .name(JavaName.fromString("surname"))
            .description("desc")
            .javaType(stringType().withNullability(NULLABLE))
            .necessity(Necessity.REQUIRED)
            .type(JavaPojoMember.MemberType.OBJECT_MEMBER)
            .memberXml(JavaPojoMemberXml.noDefinition())
            .build();

    final JavaPojoMember nameMember =
        javaPojoMemberBuilder()
            .pojoName(invoiceName())
            .name(JavaName.fromString("name"))
            .description("desc")
            .javaType(stringType().withNullability(NULLABLE))
            .necessity(Necessity.OPTIONAL)
            .type(JavaPojoMember.MemberType.OBJECT_MEMBER)
            .memberXml(JavaPojoMemberXml.noDefinition())
            .build();

    final JavaObjectPojo pojo = JavaPojos.objectPojo(PList.of(surnameMember, nameMember));

    final Generator<WitherGenerator.WitherContent, PojoSettings> generator = witherGenerator();
    final Writer writer =
        generator.generate(pojo.getWitherContent(), defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
