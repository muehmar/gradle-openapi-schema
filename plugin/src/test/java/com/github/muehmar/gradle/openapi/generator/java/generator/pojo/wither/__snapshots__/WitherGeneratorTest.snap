allNullabilityAndNecessityVariants[optionalListWithNullableItems]=[
java.util.Optional

/**
 * OptionalListWithNullableItems
 */
public NecessityAndNullabilityDto withOptionalListWithNullableItems(List<String> optionalListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}

/**
 * OptionalListWithNullableItems
 */
public NecessityAndNullabilityDto withOptionalListWithNullableItems(Optional<List<String>> optionalListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems.orElse(null),
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}

/**
 * OptionalListWithNullableItems
 */
public NecessityAndNullabilityDto withOptionalListWithNullableItems_(List<Optional<String>> optionalListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      unwrapNullableItemsList(optionalListWithNullableItems),
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}

/**
 * OptionalListWithNullableItems
 */
public NecessityAndNullabilityDto withOptionalListWithNullableItems_(Optional<List<Optional<String>>> optionalListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      unwrapNullableItemsList(optionalListWithNullableItems.orElse(null)),
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}
]


allNullabilityAndNecessityVariants[optionalNullableListWithNullableItems]=[
com.github.muehmar.openapi.util.Tristate
java.util.Optional

/**
 * OptionalNullableListWithNullableItems
 */
public NecessityAndNullabilityDto withOptionalNullableListWithNullableItems(List<String> optionalNullableListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      false,
      additionalProperties
    );
}

/**
 * OptionalNullableListWithNullableItems
 */
public NecessityAndNullabilityDto withOptionalNullableListWithNullableItems(Tristate<List<String>> optionalNullableListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems.onValue(val -> val).onNull(() -> null).onAbsent(() -> null),
      optionalNullableListWithNullableItems.onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false),
      additionalProperties
    );
}

/**
 * OptionalNullableListWithNullableItems
 */
public NecessityAndNullabilityDto withOptionalNullableListWithNullableItems_(List<Optional<String>> optionalNullableListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      unwrapNullableItemsList(optionalNullableListWithNullableItems),
      false,
      additionalProperties
    );
}

/**
 * OptionalNullableListWithNullableItems
 */
public NecessityAndNullabilityDto withOptionalNullableListWithNullableItems_(Tristate<List<Optional<String>>> optionalNullableListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      unwrapNullableItemsList(optionalNullableListWithNullableItems.onValue(val -> val).onNull(() -> null).onAbsent(() -> null)),
      optionalNullableListWithNullableItems.onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false),
      additionalProperties
    );
}
]


allNullabilityAndNecessityVariants[optionalNullableStringVal]=[
com.github.muehmar.openapi.util.Tristate

/**
 * OptionalNullableStringVal
 */
public NecessityAndNullabilityDto withOptionalNullableStringVal(String optionalNullableStringVal) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      false,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}

/**
 * OptionalNullableStringVal
 */
public NecessityAndNullabilityDto withOptionalNullableStringVal(Tristate<String> optionalNullableStringVal) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal.onValue(val -> val).onNull(() -> null).onAbsent(() -> null),
      optionalNullableStringVal.onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false),
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}
]


allNullabilityAndNecessityVariants[optionalStringVal]=[
java.util.Optional

/**
 * OptionalStringVal
 */
public NecessityAndNullabilityDto withOptionalStringVal(String optionalStringVal) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}

/**
 * OptionalStringVal
 */
public NecessityAndNullabilityDto withOptionalStringVal(Optional<String> optionalStringVal) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal.orElse(null),
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}
]


allNullabilityAndNecessityVariants[requiredListWithNullableItems]=[
java.util.Optional

/**
 * RequiredListWithNullableItems
 */
public NecessityAndNullabilityDto withRequiredListWithNullableItems(List<String> requiredListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}

/**
 * RequiredListWithNullableItems
 */
public NecessityAndNullabilityDto withRequiredListWithNullableItems_(List<Optional<String>> requiredListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      unwrapNullableItemsList(requiredListWithNullableItems),
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}
]


allNullabilityAndNecessityVariants[requiredNullableListWithNullableItems]=[
java.util.Optional

/**
 * RequiredNullableListWithNullableItems
 */
public NecessityAndNullabilityDto withRequiredNullableListWithNullableItems(List<String> requiredNullableListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      true,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}

/**
 * RequiredNullableListWithNullableItems
 */
public NecessityAndNullabilityDto withRequiredNullableListWithNullableItems(Optional<List<String>> requiredNullableListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems.orElse(null),
      true,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}

/**
 * RequiredNullableListWithNullableItems
 */
public NecessityAndNullabilityDto withRequiredNullableListWithNullableItems_(List<Optional<String>> requiredNullableListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      unwrapNullableItemsList(requiredNullableListWithNullableItems),
      true,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}

/**
 * RequiredNullableListWithNullableItems
 */
public NecessityAndNullabilityDto withRequiredNullableListWithNullableItems_(Optional<List<Optional<String>>> requiredNullableListWithNullableItems) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      unwrapNullableItemsList(requiredNullableListWithNullableItems.orElse(null)),
      true,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}
]


allNullabilityAndNecessityVariants[requiredNullableStringVal]=[
java.util.Optional

/**
 * RequiredNullableStringVal
 */
public NecessityAndNullabilityDto withRequiredNullableStringVal(String requiredNullableStringVal) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      true,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}

/**
 * RequiredNullableStringVal
 */
public NecessityAndNullabilityDto withRequiredNullableStringVal(Optional<String> requiredNullableStringVal) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal.orElse(null),
      true,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}
]


allNullabilityAndNecessityVariants[requiredStringVal]=[
.
.
/**
 * RequiredStringVal
 */
public NecessityAndNullabilityDto withRequiredStringVal(String requiredStringVal) {
  return
    new NecessityAndNullabilityDto(
      requiredStringVal,
      requiredNullableStringVal,
      isRequiredNullableStringValPresent,
      optionalStringVal,
      isOptionalStringValNotNull,
      optionalNullableStringVal,
      isOptionalNullableStringValNull,
      requiredListWithNullableItems,
      requiredNullableListWithNullableItems,
      isRequiredNullableListWithNullableItemsPresent,
      optionalListWithNullableItems,
      isOptionalListWithNullableItemsNotNull,
      optionalNullableListWithNullableItems,
      isOptionalNullableListWithNullableItemsNull,
      additionalProperties
    );
}
]


illegalIdentifierPojo=[
com.github.muehmar.openapi.util.Tristate

/**
 * Switch
 */
public Illegal_IdentifierDto withSwitch(String switch_) {
  return
    new Illegal_IdentifierDto(
      switch_,
      false,
      point_,
      isPoint_Null,
      additionalProperties
    );
}

/**
 * Switch
 */
public Illegal_IdentifierDto withSwitch(Tristate<String> switch_) {
  return
    new Illegal_IdentifierDto(
      switch_.onValue(val -> val).onNull(() -> null).onAbsent(() -> null),
      switch_.onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false),
      point_,
      isPoint_Null,
      additionalProperties
    );
}

/**
 * Point
 */
public Illegal_IdentifierDto withPoint_(String point_) {
  return
    new Illegal_IdentifierDto(
      switch_,
      isSwitchNull,
      point_,
      false,
      additionalProperties
    );
}

/**
 * Point
 */
public Illegal_IdentifierDto withPoint_(Tristate<String> point_) {
  return
    new Illegal_IdentifierDto(
      switch_,
      isSwitchNull,
      point_.onValue(val -> val).onNull(() -> null).onAbsent(() -> null),
      point_.onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false),
      additionalProperties
    );
}
]


noAdditionalProperties=[
.
.
/**
 * stringVal
 */
public ObjectDto withStringVal(String stringVal) {
  return
    new ObjectDto(
      stringVal,
      birthdate
    );
}
]


propertyNameMatchesSubstringOfOtherProperty=[
com.github.muehmar.openapi.util.Tristate
java.util.Optional

/**
 * desc
 */
public ObjectPojo1Dto withSurname(String surname) {
  return
    new ObjectPojo1Dto(
      surname,
      true,
      name,
      isNameNull,
      additionalProperties
    );
}

/**
 * desc
 */
public ObjectPojo1Dto withSurname(Optional<String> surname) {
  return
    new ObjectPojo1Dto(
      surname.orElse(null),
      true,
      name,
      isNameNull,
      additionalProperties
    );
}

/**
 * desc
 */
public ObjectPojo1Dto withName(String name) {
  return
    new ObjectPojo1Dto(
      surname,
      isSurnamePresent,
      name,
      false,
      additionalProperties
    );
}

/**
 * desc
 */
public ObjectPojo1Dto withName(Tristate<String> name) {
  return
    new ObjectPojo1Dto(
      surname,
      isSurnamePresent,
      name.onValue(val -> val).onNull(() -> null).onAbsent(() -> null),
      name.onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false),
      additionalProperties
    );
}
]